carousel/core/calculators.py

```diff
@@ -107,101 +102,107 @@ class Calculator(object):
+            except (KeyError, TypeError):
+                a = outargs[a]
+                avar = outvar[a]
+            else:
+                avar = datvar[a]
+            for n in xrange(argn):
+                b = vargs[n]
+                try:
+                    b = datargs[b]
+                except (KeyError, TypeError):
+                    b = outargs[b]
+                c = avar.get(b, 0.0)
+                try:
+                    nobs = max(nobs, len(c))
+                except (TypeError, ValueError):
+                    LOGGER.debug('c of %s vs %s = %g', a, b, c)
+        # covariance matrix is initially zeros
+        cov = np.zeros((nobs, argn, argn))
+        # loop over arguments in both directions, fill in covariance
+        for m in xrange(argn):
+            a = vargs[m]
+            try:
+                a = datargs[a]
+            except (KeyError, TypeError):
+                a = outargs[a]
+                avar = outvar[a]
+            else:
+                avar = datvar[a]
+            for n in xrange(argn):
+                b = vargs[n]
+                try:
+                    b = datargs[b]
+                except (KeyError, TypeError):
+                    b = outargs[b]
+                cov[:, m, n] = avar.get(b, 0.0)
+        if nobs == 1:
+            cov = cov.squeeze()  # squeeze out any extra dimensions
+        LOGGER.debug('covariance:\n%r', cov)
+        return cov
 
-    def calculate(self, calc, formula_reg, data_reg, out_reg, timestep):
+    def calculate(self, calc, formula_reg, data_reg, out_reg,
+                  timestep=None, idx=None):
         """
-        Execute static calculations.
+        Execute calculation
 
+        :param calc: calculation, with formula, args and return keys
+        :type calc: dict
         :param formula_reg: Registry of formulas.
         :type formula_reg: :class:`~carousel.core.FormulaRegistry`
         :param data_reg: Data registry.
-        :type data_reg: \
-            :class:`~carousel.core.data_sources.DataRegistry`
+        :type data_reg: :class:`~carousel.core.data_sources.DataRegistry`
         :param out_reg: Outputs registry.
         :type out_reg: :class:`~carousel.core.outputs.OutputRegistry`
-        :param timestep: simulation interval length [time]
+        :param timestep: simulation interval length [time], default is ``None``
+        :param idx: interval index, default is ``None``
+        :type idx: int
         """
         # get the formula-key from each static calc
         formula = calc['formula']  # name of formula in calculation
         func = formula_reg[formula]  # formula function object
-        args = calc['args']  # calculation arguments
-        # separate data and output arguments
-        datargs, outargs = args.get('data', []), args.get('outputs', [])
         fargs = formula_reg.args.get(formula, [])  # formula arguments
         constants = formula_reg.isconstant.get(formula)  # constant args
```

Comment:
Ok, so it looks like things got moved around and I understand that part.
I can't say that I really understand everything that's happening in terms of the interactions with the registry and the covariance calculation, but that's going to be for later otherwise I'll be even slower than I already am.